In [1]:
import pandas as pd

#Import classes
from datatransform import DataTransform
from dataframeinfo import DataFrameInfo
from plotter import Plotter
from dataframetransform import DataFrameTransform


ModuleNotFoundError: No module named 'missingno'

In [ ]:
#Creating class instances
data_trans = DataTransform()
data_info = DataFrameInfo()
plot = Plotter()
df_trans = DataFrameTransform()

In [ ]:
df = pd.read_csv('loan_payments.csv', index_col='id')       #load data from local csv file
pd.set_option('display.max_columns', None)      #to display all columns
df

member_id  loan_amount  funded_amount  funded_amount_inv       term  \
id                                                                              
38676116   41461848         8000         8000.0             8000.0  36 months   
38656203   41440010        13200        13200.0            13200.0  36 months   
38656154   41439961        16000        16000.0            16000.0  36 months   
38656128   41439934        15000        15000.0            15000.0  36 months   
38656121   41439927        15000        15000.0            15000.0  36 months   
...             ...          ...            ...                ...        ...   
76597         76583         5000         5000.0             1775.0  36 months   
142608        74724         5000         5000.0             2350.0  36 months   
117045        70978         3500         3500.0             2225.0  36 months   
88854         70699         5000         5000.0              225.0  36 months   
72323         70694         5000         5000.0              350.0  36 months   

          int_rate  instalment grade sub_grade employment_length  \
id                                                                 
38676116      7.49      248.82     A        A4           5 years   
38656203      6.99      407.52     A        A3           9 years   
38656154      7.49      497.63     A        A4           8 years   
38656128     14.31      514.93     C        C4            1 year   
38656121      6.03      456.54     A        A1         10+ years   
...            ...         ...   ...       ...               ...   
76597         9.01      159.03     B        B2            1 year   
142608       10.59      162.73     C        C2          < 1 year   
117045        7.43      108.77     A        A2         10+ years   
88854         7.43      155.38     A        A2           4 years   
72323        11.54      164.98     C        C5           9 years   

         home_ownership  annual_inc verification_status issue_date  \
id                                                                   
38676116       MORTGAGE     46000.0        Not Verified   Jan-2021   
38656203           RENT     50000.0        Not Verified   Jan-2021   
38656154       MORTGAGE     73913.0     Source Verified   Jan-2021   
38656128           RENT     42000.0     Source Verified   Jan-2021   
38656121       MORTGAGE    145000.0            Verified   Jan-2021   
...                 ...         ...                 ...        ...   
76597          MORTGAGE    250000.0        Not Verified   Jul-2013   
142608             RENT     15000.0        Not Verified   Oct-2013   
117045         MORTGAGE    300000.0        Not Verified   Aug-2013   
88854              RENT    200000.0        Not Verified   Aug-2013   
72323          MORTGAGE     55000.0        Not Verified   Jul-2013   

                                                loan_status payment_plan  \
id                                                                         
38676116                                            Current            n   
38656203                                            Current            n   
38656154                                         Fully Paid            n   
38656128                                         Fully Paid            n   
38656121                                            Current            n   
...                                                     ...          ...   
76597                                            Fully Paid            n   
142608                                           Fully Paid            n   
117045    Does not meet the credit policy. Status:Fully ...            n   
88854                                            Fully Paid            n   
72323     Does not meet the credit policy. Status:Charge...            n   

                     purpose    dti  delinq_2yrs earliest_credit_line  \
id                                                                      
38676116         credit_c

In [ ]:
df = df.convert_dtypes()
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 54231 entries, 38676116 to 72323
Data columns (total 42 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   member_id                    54231 non-null  Int64  
 1   loan_amount                  54231 non-null  Int64  
 2   funded_amount                51224 non-null  Int64  
 3   funded_amount_inv            54231 non-null  Float64
 4   term                         49459 non-null  string 
 5   int_rate                     49062 non-null  Float64
 6   instalment                   54231 non-null  Float64
 7   grade                        54231 non-null  string 
 8   sub_grade                    54231 non-null  string 
 9   employment_length            52113 non-null  string 
 10  home_ownership               54231 non-null  string 
 11  annual_inc                   54231 non-null  Float64
 12  verification_status          54231 non-null  string 
 13  issue_date    

In [ ]:
#Remove "Source" from column to simplify column

df['verification_status'] = df['verification_status'].str.replace('Source', '').str.strip()

#Changes string columns to period(m)
date_columns = ['issue_date', 'earliest_credit_line', 'last_payment_date', 'next_payment_date','last_credit_pull_date']
for column in date_columns:
    data_trans.string_to_date(df, column)

df

/Users/ssalih/EDA/datatransform.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[date_column] = pd.to_datetime(df[date_column], errors='coerce').dt.to_period('M')
/Users/ssalih/EDA/datatransform.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[date_column] = pd.to_datetime(df[date_column], errors='coerce').dt.to_period('M')
/Users/ssalih/EDA/datatransform.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[date_column] = pd.to_datetime(df[date_column], errors='coerce').dt.to_period('M')
/Users/ssalih/EDA/datatransform.py:8: UserWarning: Could not infer format

member_id  loan_amount  funded_amount  funded_amount_inv       term  \
id                                                                              
38676116   41461848         8000           8000             8000.0  36 months   
38656203   41440010        13200          13200            13200.0  36 months   
38656154   41439961        16000          16000            16000.0  36 months   
38656128   41439934        15000          15000            15000.0  36 months   
38656121   41439927        15000          15000            15000.0  36 months   
...             ...          ...            ...                ...        ...   
76597         76583         5000           5000             1775.0  36 months   
142608        74724         5000           5000             2350.0  36 months   
117045        70978         3500           3500             2225.0  36 months   
88854         70699         5000           5000              225.0  36 months   
72323         70694         5000           5000              350.0  36 months   

          int_rate  instalment grade sub_grade employment_length  \
id                                                                 
38676116      7.49      248.82     A        A4           5 years   
38656203      6.99      407.52     A        A3           9 years   
38656154      7.49      497.63     A        A4           8 years   
38656128     14.31      514.93     C        C4            1 year   
38656121      6.03      456.54     A        A1         10+ years   
...            ...         ...   ...       ...               ...   
76597         9.01      159.03     B        B2            1 year   
142608       10.59      162.73     C        C2          < 1 year   
117045        7.43      108.77     A        A2         10+ years   
88854         7.43      155.38     A        A2           4 years   
72323        11.54      164.98     C        C5           9 years   

         home_ownership  annual_inc verification_status issue_date  \
id                                                                   
38676116       MORTGAGE     46000.0        Not Verified    2021-01   
38656203           RENT     50000.0        Not Verified    2021-01   
38656154       MORTGAGE     73913.0            Verified    2021-01   
38656128           RENT     42000.0            Verified    2021-01   
38656121       MORTGAGE    145000.0            Verified    2021-01   
...                 ...         ...                 ...        ...   
76597          MORTGAGE    250000.0        Not Verified    2013-07   
142608             RENT     15000.0        Not Verified    2013-10   
117045         MORTGAGE    300000.0        Not Verified    2013-08   
88854              RENT    200000.0        Not Verified    2013-08   
72323          MORTGAGE     55000.0        Not Verified    2013-07   

                                                loan_status payment_plan  \
id                                                                         
38676116                                            Current            n   
38656203                                            Current            n   
38656154                                         Fully Paid            n   
38656128                                         Fully Paid            n   
38656121                                            Current            n   
...                                                     ...          ...   
76597                                            Fully Paid            n   
142608                                           Fully Paid            n   
117045    Does not meet the credit policy. Status:Fully ...            n   
88854                                            Fully Paid            n   
72323     Does not meet the credit policy. Status:Charge...            n   

                     purpose    dti  delinq_2yrs earliest_credit_line  \
id                                                                      
38676116         credit_c

In [ ]:
#Averages testing

# mean_value = data_info.mean(df, 'loan_amount')
# mean_value

mean_value = data_info.mean(df)
mean_value

median_value = data_info.median(df)
median_value

std_value = data_info.std(df)
std_value

member_id                      10312806.882093
loan_amount                        8082.196709
funded_amount                      8019.017599
funded_amount_inv                  8099.473527
int_rate                              4.392893
instalment                          238.920012
annual_inc                        51589.339577
dti                                   7.623124
delinq_2yrs                           0.706706
inq_last_6mths                        1.261421
mths_since_last_delinq                21.92324
mths_since_last_record               36.137264
open_accounts                         4.797169
total_accounts                       11.587025
out_prncp                          5154.893646
out_prncp_inv                      5153.043643
total_payment                      8391.043845
total_payment_inv                  8363.508506
total_rec_prncp                    6958.124264
total_rec_int                      2581.657345
total_rec_late_fee                    6.215792
recoveries   

In [ ]:
#Counting distinct values

distinct_values = data_info.distinct_count(df, 'grade')
distinct_values



# category_columns = df.select_dtypes(include=['category']).columns.tolist()

# print("Columns with 'category' dtype:", category_columns)


7

In [ ]:
# Print out the shape of the DataFrame
df_shape = data_info.shape_print(df)
df_shape

(54231, 42)

In [ ]:
# Generate a count/percentage count of NULL values in each column

total_null = data_info.null_count(df)
total_null

percentage_null = data_info.null_percentage(df)
percentage_null


member_id                      0.000000
loan_amount                    0.000000
funded_amount                  0.055448
funded_amount_inv              0.000000
term                           0.087994
int_rate                       0.095314
instalment                     0.000000
grade                          0.000000
sub_grade                      0.000000
employment_length              0.039055
home_ownership                 0.000000
annual_inc                     0.000000
verification_status            0.000000
issue_date                     0.000000
loan_status                    0.000000
payment_plan                   0.000000
purpose                        0.000000
dti                            0.000000
delinq_2yrs                    0.000000
earliest_credit_line           0.000000
inq_last_6mths                 0.000000
mths_since_last_delinq         0.571666
mths_since_last_record         0.886025
open_accounts                  0.000000
total_accounts                 0.000000


#### Based on the above we can see the following have a signficantly large null count/ percentage:

mths_since_last_delinq             (0.571666)  
mths_since_last_record             (0.886025)  
next_payment_date                  (0.601280)  
mths_since_last_major_derog        (0.861721)  

As such those will be dropped

In [ ]:
columns_to_drop = ['mths_since_last_delinq', 'mths_since_last_record', 'next_payment_date', 'mths_since_last_major_derog']

for column in columns_to_drop:
    df_trans.drop_column(df,column)



df.shape
df


member_id  loan_amount  funded_amount  funded_amount_inv       term  \
id                                                                              
38676116   41461848         8000           8000             8000.0  36 months   
38656203   41440010        13200          13200            13200.0  36 months   
38656154   41439961        16000          16000            16000.0  36 months   
38656128   41439934        15000          15000            15000.0  36 months   
38656121   41439927        15000          15000            15000.0  36 months   
...             ...          ...            ...                ...        ...   
76597         76583         5000           5000             1775.0  36 months   
142608        74724         5000           5000             2350.0  36 months   
117045        70978         3500           3500             2225.0  36 months   
88854         70699         5000           5000              225.0  36 months   
72323         70694         5000           5000              350.0  36 months   

          int_rate  instalment grade sub_grade employment_length  \
id                                                                 
38676116      7.49      248.82     A        A4           5 years   
38656203      6.99      407.52     A        A3           9 years   
38656154      7.49      497.63     A        A4           8 years   
38656128     14.31      514.93     C        C4            1 year   
38656121      6.03      456.54     A        A1         10+ years   
...            ...         ...   ...       ...               ...   
76597         9.01      159.03     B        B2            1 year   
142608       10.59      162.73     C        C2          < 1 year   
117045        7.43      108.77     A        A2         10+ years   
88854         7.43      155.38     A        A2           4 years   
72323        11.54      164.98     C        C5           9 years   

         home_ownership  annual_inc verification_status issue_date  \
id                                                                   
38676116       MORTGAGE     46000.0        Not Verified    2021-01   
38656203           RENT     50000.0        Not Verified    2021-01   
38656154       MORTGAGE     73913.0            Verified    2021-01   
38656128           RENT     42000.0            Verified    2021-01   
38656121       MORTGAGE    145000.0            Verified    2021-01   
...                 ...         ...                 ...        ...   
76597          MORTGAGE    250000.0        Not Verified    2013-07   
142608             RENT     15000.0        Not Verified    2013-10   
117045         MORTGAGE    300000.0        Not Verified    2013-08   
88854              RENT    200000.0        Not Verified    2013-08   
72323          MORTGAGE     55000.0        Not Verified    2013-07   

                                                loan_status payment_plan  \
id                                                                         
38676116                                            Current            n   
38656203                                            Current            n   
38656154                                         Fully Paid            n   
38656128                                         Fully Paid            n   
38656121                                            Current            n   
...                                                     ...          ...   
76597                                            Fully Paid            n   
142608                                           Fully Paid            n   
117045    Does not meet the credit policy. Status:Fully ...            n   
88854                                            Fully Paid            n   
72323     Does not meet the credit policy. Status:Charge...            n   

                     purpose    dti  delinq_2yrs earliest_credit_line  \
id                                                                      
38676116         credit_c

In [ ]:
# If the percentage of missing data is relatively small, you may choose to impute the missing 
# values using statistical methods such as mean or median imputation. 

percentage_null = data_info.null_percentage(df)
percentage_null



member_id                     0.000000
loan_amount                   0.000000
funded_amount                 0.055448
funded_amount_inv             0.000000
term                          0.087994
int_rate                      0.095314
instalment                    0.000000
grade                         0.000000
sub_grade                     0.000000
employment_length             0.039055
home_ownership                0.000000
annual_inc                    0.000000
verification_status           0.000000
issue_date                    0.000000
loan_status                   0.000000
payment_plan                  0.000000
purpose                       0.000000
dti                           0.000000
delinq_2yrs                   0.000000
earliest_credit_line          0.000000
inq_last_6mths                0.000000
open_accounts                 0.000000
total_accounts                0.000000
out_prncp                     0.000000
out_prncp_inv                 0.000000
total_payment            

In [ ]:
cat_impute = df_trans.impute_categorical_nulls(df)
cat_impute

percentage_null = data_info.null_percentage(df)
percentage_null

# df.dtypes

member_id                     0.000000
loan_amount                   0.000000
funded_amount                 0.055448
funded_amount_inv             0.000000
term                          0.087994
int_rate                      0.095314
instalment                    0.000000
grade                         0.000000
sub_grade                     0.000000
employment_length             0.039055
home_ownership                0.000000
annual_inc                    0.000000
verification_status           0.000000
issue_date                    0.000000
loan_status                   0.000000
payment_plan                  0.000000
purpose                       0.000000
dti                           0.000000
delinq_2yrs                   0.000000
earliest_credit_line          0.000000
inq_last_6mths                0.000000
open_accounts                 0.000000
total_accounts                0.000000
out_prncp                     0.000000
out_prncp_inv                 0.000000
total_payment            


funded_amount                 0.055448
term                          0.087994
int_rate                      0.095314
employment_length             0.039055
last_payment_date             0.001346
last_credit_pull_date         0.000129
collections_12_mths_ex_med    0.000940

dtype: float64

If data null data column count is < 1% (ie 0.01) this suggests that this make up a small subset of the data these rows can be removed.

In [ ]:
# small_null_percentage_check = ['funded_amount', 'term', 'int_rate', 'employment_length', 'last_payment_date','last_credit_pull_date', 'collections_12_mths_ex_med ']

# drops_null_rows_under_threshold = df_trans.drop_rows_with_sparse_nulls(df)

# percentage_null

sorted_null_percentage = percentage_null.sort_values(ascending=False)
print(sorted_null_percentage)

sparse_null_columns = df_trans.columns_with_sparse_nulls(df)
print(sparse_null_columns)





int_rate                      0.095314
term                          0.087994
funded_amount                 0.055448
employment_length             0.039055
last_payment_date             0.001346
collections_12_mths_ex_med    0.000940
last_credit_pull_date         0.000129
total_rec_int                 0.000000
out_prncp                     0.000000
out_prncp_inv                 0.000000
total_payment                 0.000000
total_payment_inv             0.000000
total_rec_prncp               0.000000
member_id                     0.000000
total_rec_late_fee            0.000000
recoveries                    0.000000
collection_recovery_fee       0.000000
open_accounts                 0.000000
last_payment_amount           0.000000
policy_code                   0.000000
total_accounts                0.000000
earliest_credit_line          0.000000
inq_last_6mths                0.000000
loan_amount                   0.000000
delinq_2yrs                   0.000000
dti                      

In [ ]:
# df.dropna(subset=sparse_null_columns)

# percentage_null

# Check for null values before dropping rows
print("Null values before dropping rows:")
print(df[sparse_null_columns].isnull().sum())

# Drop rows containing null values in specified columns
df = df.dropna(subset=sparse_null_columns)

# Check for null values after dropping rows
print("\nNull values after dropping rows:")
print(df[sparse_null_columns].isnull().sum())

percentage_null = data_info.null_percentage(df)
percentage_null


Null values before dropping rows:
last_payment_date             73
last_credit_pull_date          7
collections_12_mths_ex_med    51
dtype: int64

Null values after dropping rows:
last_payment_date             0
last_credit_pull_date         0
collections_12_mths_ex_med    0
dtype: int64


member_id                     0.000000
loan_amount                   0.000000
funded_amount                 0.055416
funded_amount_inv             0.000000
term                          0.088022
int_rate                      0.095360
instalment                    0.000000
grade                         0.000000
sub_grade                     0.000000
employment_length             0.039076
home_ownership                0.000000
annual_inc                    0.000000
verification_status           0.000000
issue_date                    0.000000
loan_status                   0.000000
payment_plan                  0.000000
purpose                       0.000000
dti                           0.000000
delinq_2yrs                   0.000000
earliest_credit_line          0.000000
inq_last_6mths                0.000000
open_accounts                 0.000000
total_accounts                0.000000
out_prncp                     0.000000
out_prncp_inv                 0.000000
total_payment            

### Now there are 3 columns left 
lets figure out how the data is missing? MCAR?
Need to use Plotter


In [ ]:
plot.missing_matrix(df)

NameError: name 'plot' is not defined